components/script/dom/elementinternals.rs

```diff
@@ -0,0 +1,371 @@
+            .upcast::<Element>()
+            .get_string_attribute(&local_name!("name"));
+        if name == "" {
+            return;
+        }
+        match &*self.submission_value.borrow() {
+            SubmissionValue::FormData(_) => unreachable!(),
+            SubmissionValue::None => {},
+            SubmissionValue::USVString(s) => {
+                entry_list.push(FormDatum {
+                    ty: DOMString::from("string"),
+                    name: name,
+                    value: FormDatumValue::String(DOMString::from(s.to_string())),
+                });
+            },
+            SubmissionValue::File(f) => {
+                entry_list.push(FormDatum {
+                    ty: DOMString::from("file"),
+                    name: name,
+                    value: FormDatumValue::File(DomRoot::from_ref(&*f)),
+                });
+            },
+        }
+    }
+}
+
+impl ElementInternalsMethods for ElementInternals {
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-setformvalue>
+    fn SetFormValue(
+        &self,
+        value: Option<FileOrUSVStringOrFormData>,
+        maybe_state: Option<Option<FileOrUSVStringOrFormData>>,
+    ) -> ErrorResult {
+        // Steps 1-2: If element is not a form-associated custom element, then throw a "NotSupportedError" DOMException
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+
+        // Step 3: Set target element's submission value
+        self.set_submission_value(submission_value_from(&value));
+
+        match maybe_state {
+            // Step 4: If the state argument of the function is omitted, set element's state to its submission value
+            None => self.set_state(submission_value_from(&value)),
+            // Steps 5-6: Otherwise, set element's state to state
+            Some(state) => self.set_state(submission_value_from(&state)),
+        }
+        Ok(())
+    }
+
+    /// <https://html.spec.whatwg.org/multipage#dom-elementinternals-setvalidity>
+    fn SetValidity(
+        &self,
+        flags: &ValidityStateFlags,
+        message: Option<DOMString>,
+        anchor: Option<&HTMLElement>,
+    ) -> ErrorResult {
+        // Steps 1-2: Check form-associated custom element
+        if !self.is_target_form_associated() {
+            return Err(Error::NotSupported);
+        }
+
+        // Step 3: Check bits and message
+        let bits = validity_bits(flags);
+        if !bits.is_empty() && !message.iter().any(|m| m.len() > 0) {
+            return Err(Error::Type(
+                "Setting an element to invalid requires a message string as the second argument."
+                    .to_string(),
+            ));
+        }
+
+        // Step 4: For each entry flag → value of flags, set element's validity flag with the name flag to value
+        self.validity_state().update_invalid_flags(bits);
+
+        // Step 5: Set element's validation message to the empty string
```

Comment:
Truncating the text is pretty confusing here: 

```suggestion
        // Step 5: Set element's validation message  to the empty string if message is not given
        // or all of element's validity flags are false, or to message otherwise.
```